tests/auto/resources/sugarmockprotocol.cpp

```diff
@@ -0,0 +1,70 @@
+/*
+  This file is part of FatCRM, a desktop application for SugarCRM written by KDAB.
+
+  Copyright (C) 2015-2017 Klarälvdalens Datakonsult AB, a KDAB Group company, info@kdab.com
+  Authors: Jeremy Entressangle <jeremy.entressangle@kdab.com>
+
+  This program is free software; you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 2 of the License, or
+  (at your option) any later version.
+
+  This program is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "sugarmockprotocol.h"
+#include "sugarjob.h"
+
+SugarMockProtocol::SugarMockProtocol()
+    : mServerNotFound(false)
+{
+
+}
+
+int SugarMockProtocol::login(const QString &user, const QString &password, QString &sessionId, QString &errorMessage)
+{
+    if (!mServerNotFound) {
+        if (user == "user" && password == "password") {
+            sessionId = "1";
+            return KJob::NoError;
+        } else {
+            int error = SugarJob::Errors::LoginError;
+            errorMessage = QString("Login for user %1 on %2 failed: %3").arg(user).arg("test").arg(error);
+            return error;
+        }
+    } else {
+        return SugarJob::Errors::CouldNotConnectError;
+    }
+}
+
+void SugarMockProtocol::logout()
+{
+}
+
+void SugarMockProtocol::setSession(SugarSession *session)
+{
+    Q_UNUSED(session);
+}
+
+void SugarMockProtocol::getEntriesCount(const ListEntriesScope &scope, const QString &moduleName, const QString &query)
+{
+    Q_UNUSED(scope);
+    Q_UNUSED(moduleName);
+    Q_UNUSED(query);
+}
+
+void SugarMockProtocol::listEntries(const ListEntriesScope &scope, const QString &moduleName, const QString &query,
```

Comment:
This will only be useful if it's actually implemented to return something.

You could add a list of accounts, pre-initialized with 1 account, and in getEntriesCount(), if moduleName is Accounts, return 1 (well, count() of that list, to stay generic), and here ..... well, we need to return the account details, but that's where it gets interesting ;)

We could use KDSoapGenerated::TNS__Get_entry_list_result as the return type for this listEntries method. It would make things easy in the SoapProtocol implementation. Here however, the mock needs code to convert a SugarAccount to a KDSoapGenerated::TNS__Name_value_list -- that code exists in AccountsHandler::setEntry and needs to be extracted to a new virtual method in the handlers (without the call to payload<> at the beginning, and without the call to asyncSet_entry at the end, just the stuff in between).

I think this means passing the handler pointer to listEntries, too.